RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu

```diff
@@ -0,0 +1,417 @@
+#include <iostream>
+#include <limits>
+
+#include <cuda.h>
+
+#include "CondFormats/EcalObjects/interface/EcalPulseCovariances.h"
+#include "CondFormats/EcalObjects/interface/EcalPulseShapes.h"
+#include "DataFormats/EcalDigi/interface/EcalDataFrame.h"
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+#include "DataFormats/Math/interface/approx_exp.h"
+#include "DataFormats/Math/interface/approx_log.h"
+
+#include "AmplitudeComputationCommonKernels.h"
+#include "AmplitudeComputationKernels.h"
+#include "KernelHelpers.h"
+
+namespace ecal {
+  namespace multifit {
+
+    void eigen_solve_submatrix(SampleMatrix& mat, SampleVector& invec, SampleVector& outvec, unsigned NP) {
+      using namespace Eigen;
+      switch (NP) {  // pulse matrix is always square.
+        case 10: {
+          Matrix<SampleMatrix::Scalar, 10, 10> temp = mat.topLeftCorner<10, 10>();
```

Comment:
There is a lot of code duplication here, can you think of a way to reduce this? Why not something like `eigen_solve_submatrix<NP>`?

Reply:
I assume this could be simplified to
```c++
outvec.head<10>() = mat.topLeftCorner<10, 10>().ldlt().solve(invec.head<10>());
```

Then, would moving that to a template function really ave much ?